### PR TITLE
Unify restRequest with Backbone.ajax and $.ajax

### PR DESCRIPTION
**Pull request by brianhelba**
Now, vanilla Backbone Models can use `Backbone.sync`-calling methods with the Girder API. The call chain will typically be:
`Backbone.sync` -> `Backbone.ajax` -> `restRequest` -> `Backbone.$.ajax` -> `$.ajax`

`restRequest` will now expect a strict superset (just adding `girderToken`) of the parameters that `$.ajax` expects:
* Existing calls to `restRequest` have been updated to use the `url` and `method` arguments, as these are preferred by `$.ajax`, in place of `path` and `type` respectively. Use of the `url` argument is now deprecated, and will print a warning. Use of the `type` argument is also deprecated, but will not print anything, because it may still be used by Backbone.
* The `dataType` argument to `restRequest` will no longer be set to `'json'` by default . Instead, the default behavior of `$.ajax` is allowed to occur, autodetecting the response type from MIME headers,
which Girder's API will set to `'application/json'` when appropriate.

The `mockRestRequest` and `unmockRestRequest` methods are now deprecated, as mocking code for tests should not contaminate the functionality within source files. The tests using these functions have been updated to use Jasmine spies.

This PR's commits are structured to have one logical change per commit, for ease of review.

### CODE REVIEW

**Comment by danlamanna** [comment]
+1

It's nice to have an interface closer to $.ajax and to configure
Backbone in a way that sync will correctly work within the context of
Girder. Do you have any thoughts about what follows this PR,
particularly with regards to Model.js and what the path forward is. For
example being able to rely on Backbone.Model.save instead of our own.

**Comment by brianhelba** [comment]
@danlamanna Yes, my plan is to continue to move Girder towards upstream Backbone.

Models will probably be less controversial to unify on the base class, as we just need to ensure that all methods are triggering the correct native Backbone events. Relatedly, more work will also need to be done to move nearly all usages of `restRequest` out of Views and into Models or Collections, and ensure that these methods return Promises (testing this is going to require some updates to our test infrastructure, to better support actual unit tests). Finally, Views should be updated to rely on Promises for single async events, and native Backbone events on models otherwise.

Collections will likely require fewer lines of code to be refactored, but is going to require a very disruptive change to the `fetch` method. I don't have a well-formed plan on how to do this yet, but the least-painful way may be to make an alias to the existing `fetch`, called something like `load` (better names are welcome), find/replace all usages, and deprecate `fetch` until v3. Upon v3, we can remove Girder's implementation of `fetch`, and allow the native Backbone implementation to be exposed. Of course, we also need to do a thorough review of the native helper methods, to ensure they're doing what the Girder API expects. Again, real unit testing will be critical to this process.

Beyond https://github.com/girder/girder/issues/2030 , I haven't laid much of this out in a GH issue yet, partly because I haven't fully planned it out myself yet. If anyone wants to continue discussing this though, I'll make an issue, so this PR doesn't go too far off track.

File: clients/web/src/rest.js [hunk 1]
```diff
@@ -84,19 +84,25 @@ setStaticRoot(
...
+ * @param {Object} opts Options for the request, most of which will be passed through to `$.ajax`.
+ * @param {string} opts.url The resource path, e.g. "user/login"
+ * @param {string} [opts.method='GET'] The HTTP method to invoke.
+ * @param [opts.data] The query string or form parameter object.
+ * @param {?Function} [opts.error] An error callback, as documented in
+ *        `$.ajax <http://api.jquery.com/jQuery.ajax/>`_, or null. If not provided, this will have a
+ *        default behavior of triggering a `'g:alert'` global event, with details of the error.
+ * @param {string} [opts.girderToken] An alternative auth token to use for this request.
+ * @returns {$.Promise} A jqXHR promise, which resolves and rejects
+ *          `as documented by $.ajax<http://api.jquery.com/jQuery.ajax/#jqXHR>`_.
```
Comment (jbeezley) [inline clients/web/src/rest.js]: I don't know of any way to parse restructured text with esdoc (or jsdoc); however, there are plugins that support markdown.  In this case, you can just use the [`@url`](http://usejsdoc.org/tags-inline-link.html) tag.

File: clients/web/src/rest.js [hunk 2]
```diff
@@ -84,19 +84,25 @@ setStaticRoot(
...
  * e.g. logging in, registering, or generally filling out forms.
  *
- * @param path The resource path, e.g. "user/login"
- * @param data The form parameter object.
- * @param [type='GET'] The HTTP method to invoke.
- * @param [girderToken] An alternative auth token to use for this request.
+ * @param {Object} opts Options for the request, most of which will be passed through to `$.ajax`.
+ * @param {string} opts.url The resource path, e.g. "user/login"
+ * @param {string} [opts.method='GET'] The HTTP method to invoke.
+ * @param [opts.data] The query string or form parameter object.
```
Comment (jbeezley) [inline clients/web/src/rest.js]: Could add `{string|Object}` here?

**Comment by brianhelba** [comment]
Improved docs, per feedback from @jbeezley .

**Comment by brianhelba** [comment]
@girder/developers I'm reconsidering the rename from `path` -> `url` in this PR. While `url` is the option that `$.ajax` uses, that does not transform the URLs in any way. `restRequest` instead demands a "relative-ish" URL component, and unconditionally prepends the API root. The function documentation now clearly states the expected format of this option, but I'm not sure what is more "obvious" to a user transitioning from `$.ajax` to `restRequest`. I don't want to support both (unless it's via deprecation), as we should have *"one right way"*. Thoughts?

**Comment by jbeezley** [comment]
I was going to bring that up, but after thinking about it, I decided what you've done is okay at least for now.  `$.ajax` supports both `path` and `url`, so I don't think that distinction is relevant.  The only reason I can see the custom behavior as being a problem is with an external application that combines different API servers.  Even then, you could easily get around it by wrapping our custom `Backbone.ajax`.

A plausible alternative would be to prefix path in the base `Model.url` and `Collection.url` methods, but I would want to do that as a breaking change.

**Comment by brianhelba** [comment]
In response to some inquiries, let me elaborate a bit more on the rationale for migrating the parameter `path` -> `url`:

Lots of client code was abusing the fact that `restRequest` passed any unhandled / undocumented arguments through to `$.ajax`. It made sense to make it official that `restRequest` be a wrapper around `$.ajax` with a strict superset of parameters (to just handle Girder-specific authentication), because:
* `$.ajax` is already quite well known in the web development community, with lots of official and unofficial (StackOverflow, etc.) documentation. Adopting its API for `restRequest` (which is one of the most fundamental Girder-specific functions) should reduce barriers to entry for new developers.
* Backbone can use `restRequest` as the handler for `Backbone.ajax` without any compatibility concerns.
* Per the principle of least surprise, it did not make sense to expose another brand-new parameter (`path`) which serves essentially the same function as `url`. While there is a subtle difference in the way this parameter is handled by `restRequest` (as the API root is prepended), this is now well documented.
* Passing contradictory or redundant parameters (e.g. previously passing both `url` and `path`) can now more often be handled by `$.ajax`, leaving less custom logic in Girder.

**Pull request merged**